File: AppSets-iOS/AppSets/com/app/appsets/im/MessageToInfo.swift
//
//  MessageToInfo.swift
//  AppSets
//
//  Created by caiju Xu on 2024/7/9.
//

import Foundation

struct MessageToInfo: Bio, Equatable {
    static func == (lhs: MessageToInfo, rhs: MessageToInfo) -> Bool {
        return lhs.toType == rhs.toType &&
        lhs.id == rhs.id &&
        lhs.name == rhs.name &&
        lhs.iconUrl == rhs.iconUrl &&
        lhs.roles == rhs.roles
    }
    
    
    let toType:String
    let id: String
    var name: String? = nil
    var iconUrl:String? = nil
    var roles:String? = nil
    
    var bioUrl: String? = nil
    
    init(toType:String, id:String, name:String?, iconUrl:String?, roles:String?){
        self.toType = toType
        self.id = id
        self.name = name
        self.iconUrl = iconUrl
        self.roles = roles
    }
    
    static func fromImObj(_ imObj: any ImObj) -> MessageToInfo {
        
        let bio = imObj.bio
        
        var iconUrl:String? = nil
        if bio is UserInfo {
            iconUrl = (bio as? UserInfo)?.avatarUrl
        }
        else if bio is GroupInfo {
            iconUrl = (bio as? GroupInfo)?.iconUrl
        }
        else if bio is Application {
            iconUrl = (bio as? Application)?.iconUrl
        }
        
        if imObj is ImSingle {
            return MessageToInfo(
                toType: ImMessageConstant.TYPE_O2O,
                id: imObj.id,
                name: bio.name,
                iconUrl: iconUrl,
                roles: (bio as? UserInfo)?.roles
            )
        }else {
            return MessageToInfo(
                toType: ImMessageConstant.TYPE_O2M,
                id: imObj.id,
                name: bio.name,
                iconUrl: iconUrl,
                roles: nil
            )
        }
    }
}

File: AppSets-Android/appsets/src/main/kotlin/xcj/app/appsets/im/MessageToInfo.kt
package xcj.app.appsets.im

import xcj.app.appsets.im.message.IMMessage
import xcj.app.appsets.server.model.Application
import xcj.app.appsets.server.model.GroupInfo
import xcj.app.appsets.server.model.UserInfo
import xcj.app.starter.android.util.PurpleLogger

data class MessageToInfo(
    val toType: String,
    val iconUrl: String?,
    val roles: String?,
    val id: String,
    val name: String?
) : Bio {
    override val bioId: String
        get() = id
    override val bioName: String?
        get() = name ?: id
    override var bioUrl: Any? = null

    companion object {

        private const val TAG = "MessageToInfo"

        fun fromImObj(imObj: IMObj): MessageToInfo {
            val iconUrl = when (imObj.bio) {
                is UserInfo -> imObj.bio.avatarUrl
                is GroupInfo -> imObj.bio.iconUrl
                is Application -> imObj.bio.iconUrl
                else -> imObj.bio.bioUrl
            }
            val messageToInfo = when (imObj) {
                is IMObj.IMSingle -> {
                    MessageToInfo(
                        toType = IMMessage.TYPE_O2O,
                        iconUrl = iconUrl?.toString(),
                        roles = imObj.userRoles,
                        id = imObj.bio.bioId,
                        name = imObj.bio.bioName,
                    )
                }

                is IMObj.IMGroup -> {
                    MessageToInfo(
                        toType = IMMessage.TYPE_O2M,
                        iconUrl = iconUrl?.toString(),
                        roles = null,
                        id = imObj.bio.bioId,
                        name = imObj.bio.bioName,
                    )
                }
            }
            PurpleLogger.current.d(
                TAG,
                "fromImObj, imObj:$imObj, " +
                        "bio:${imObj.bio}, " +
                        "messageToInfo:$messageToInfo"
            )
            return messageToInfo
        }

    }
}